src/expr/src/relation/mod.rs

```diff
@@ -289,16 +287,16 @@ impl MirRelationExpr {
                 }
             }
             MirRelationExpr::Get { typ, .. } => typ.clone(),
-            MirRelationExpr::Let { body, .. } => body.typ(),
-            MirRelationExpr::Project { input, outputs } => {
-                let input_typ = input.typ();
+            MirRelationExpr::Let { .. } => input_types.first().unwrap().clone(),
```

Comment:
The `visit1` method considers `MirRelationExpr::Let` to have 2 child relations whereas the `num_inputs` methods being added in this PR considers `MirRelationExpr::Let` to have 1 child.

What about making `num_inputs` consistent with `visit1` and having this method expect `input_types.len()` to be 2 when `self` is `MirRelationExpr::Let`? You'd just change `input_types.first()` to `input_types.last()`.

Reply:
My first version did that, but that made visit-based `typ` method slower than the original version with the traversal embedded, since it was then traversing the `value` sub-graph which was otherwise skipped.

I was thinking that alternatively I could make the change you are suggesting but skip traversing the `value` sub-graph in the `pre` lambda in `typ` and insert an empty `RelationType` instead. That would make the visiting logic in the transform a bit simpler.

Reply:
Inserting an empty `RelationType` as part of the `pre` lambda sounds like a great idea. I prefer that other crates not have to be aware of the `Let` special-casing.